test/metabase/server/statistics_handler_test.clj

```diff
@@ -0,0 +1,252 @@
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/requests-active)))
+          (is (prometheus-test/approx= 2 (mt/metric-value system :jetty/dispatched-total)))
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/dispatched-active)))
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/dispatched-active-max)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-waiting-max)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-total)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-waiting)))
+          (a/>!! chan-finish-handle :done)
+          (a/<!! chan-done-request)
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-waiting-max)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-total)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/async-requests-waiting)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/requests-active)))
+          (is (prometheus-test/approx= 2 (mt/metric-value system :jetty/dispatched-total)))
+          (is (prometheus-test/approx= 0 (mt/metric-value system :jetty/dispatched-active)))
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/dispatched-active-max)))
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/responses-total {:code "4xx"})))))
+      (testing "when server responds 300"
+        (with-server 300 false
+          [^LocalConnector connector chan-start-handle chan-finish-handle chan-done-request]
+          (.executeRequest connector "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n")
+          (a/>!! chan-start-handle :done)
+          (a/>!! chan-finish-handle :done)
+          (a/<!! chan-done-request)
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/responses-total {:code "3xx"})))))
+      (testing "when server responds 500"
+        (with-server 500 false
+          [^LocalConnector connector chan-start-handle chan-finish-handle chan-done-request]
+          (.executeRequest connector "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n")
+          (a/>!! chan-start-handle :done)
+          (a/>!! chan-finish-handle :done)
+          (a/<!! chan-done-request)
+          (is (prometheus-test/approx= 1 (mt/metric-value system :jetty/responses-total {:code "5xx"}))))))))
+
+(deftest test-asynchronous-metrics
+  (testing "records stats for asynchronous requests"
+    (mt/with-prometheus-system! [_ system]
+      (testing "when server responds 200"
+        (with-server 200 true
+          [^LocalConnector connector chan-start-handle chan-finish-handle chan-done-request]
+          (.executeRequest connector "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n")
+          (a/>!! chan-start-handle :done)
```

Comment:
I _think_ I understand what's going on with each channel and `>!!`/`<!!` call, but honestly it's a bit confusing.

What if `with-server` didn't return raw channels, but instead wrapped those channels into functions that you could name appropriately? So instead of `(>!! chan-start-handle :done)` you might have

```
(with-server 200 true [^LocalConnector connector start-handling! finish-handling! wait-until-request-done]
  (.executeRequest ...)
  (start-handling!)
  ;; assertions
  (finish-handling!)
  (wait-until-request-done)
  ;; assertions
  )
```

etc. Exactly the same stuff could be going on under the hood but with just a little sugar to let us not have to know e.g. "which channel do I read from, and which do I write to"?

Reply:
I didn't notice this was a backport 😭 sorry, everything looks good